src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java

```diff
@@ -219,6 +231,111 @@ public static NBTTagCompound readTag(ByteBuf from)
         }
     }
 
+    /**
+     * Write a registry entry to the stream. The serialized format is not specified and must not be relied upon.
+     * Do not use this to write to a file, it is used for client-server communication only.
+     * @param out the buffer to write to
+     * @param entry the registry entry
+     */
+    public static <T extends IForgeRegistryEntry<T>> void writeRegistryEntry(@Nonnull ByteBuf out, @Nonnull T entry)
+    {
+        FMLControlledNamespacedRegistry<T> registry = (FMLControlledNamespacedRegistry<T>)GameRegistry.findRegistry(entry.getRegistryType());
+        writeUTF8String(out, registry.getRegistryName().toString());
+        writeVarInt(out, registry.getId(entry), 5);
+    }
+
+    /**
+     * Read a registry entry from the stream. The same format as in {@link #writeRegistryEntry(ByteBuf, IForgeRegistryEntry)} is used.
+     * @param in the buffer to read from
+     * @param registry the registry the entry belongs to
+     * @return the read registry entry
+     */
+    @Nonnull
+    public static <T extends IForgeRegistryEntry<T>> T readRegistryEntry(@Nonnull ByteBuf in, @Nonnull IForgeRegistry<T> registry)
+    {
+        String registryName = readUTF8String(in);
+        int id = readVarInt(in, 5);
+        if (!registry.getRegistryName().toString().equals(registryName))
+        {
+            throw new IllegalArgumentException("Registry mismatch: " + registryName + " != " + registry.getRegistryName());
+        }
+        T thing = ((FMLControlledNamespacedRegistry<T>)registry).getRaw(id);
+        if (thing == null)
+        {
+            throw new IllegalArgumentException("Unknown ID " + id + " for registry " + registry.getRegistryName() + " received.");
+        }
+        return thing;
+    }
+
+    /**
+     * Write multiple registry entries from the same registry to the stream. The serialized format may be more compact than using
+     * {@link #writeRegistryEntry(ByteBuf, IForgeRegistryEntry)} multiple times.
+     * @param out the buffer to write to
+     * @param entries the entries to write
+     */
+    public static <T extends IForgeRegistryEntry<T>> void writeRegistryEntries(@Nonnull ByteBuf out, @Nonnull Collection<T> entries)
+    {
+        writeVarInt(out, entries.size(), 5);
+        Iterator<T> it = entries.iterator();
+        if (it.hasNext())
+        {
+            T entry = it.next();
+            FMLControlledNamespacedRegistry<T> registry = (FMLControlledNamespacedRegistry<T>)GameRegistry.findRegistry(entry.getRegistryType());
+            writeUTF8String(out, registry.getRegistryName().toString());
+            do
+            {
+                writeVarInt(out, registry.getId(entry), 5);
+                if (it.hasNext())
+                {
+                    entry = it.next();
+                }
+                else
+                {
+                    break;
+                }
+            }
+            while (true);
+        }
```

Comment:
To be neater, we can just return when the iterator has no entries, less indentation.